src/lambda/commands/deploySamApplication.ts

```diff
@@ -13,137 +13,200 @@ import * as nls from 'vscode-nls'
+
+        throwAndNotifyIfInvalid(await samCliContext.validator.detectValidSamCli())
+
+        const deployWizardResponse: SamDeployWizardResponse | undefined = await new SamDeployWizard(
+            restParams.regionProvider
+        ).run()
+
+        if (!deployWizardResponse) {
+            return
+        }
+
+        const deployParameters: DeploySamApplicationParameters = {
+            deployRootFolder: await makeTemporaryToolkitFolder('samDeploy'),
+            destinationStackName: deployWizardResponse.stackName,
+            packageBucketName: deployWizardResponse.s3Bucket,
+            parameterOverrides: deployWizardResponse.parameterOverrides,
+            profile,
+            region: deployWizardResponse.region,
+            sourceTemplatePath: deployWizardResponse.template.fsPath,
+        }
+
+        const deployApplicationPromise = buildPackageDeploy({
+            deployParameters,
+            channelLogger,
+            invoker: samCliContext.invoker,
+        }).then(async () =>
+            await del(deployParameters.deployRootFolder, {
+                force: true
+            })
+        )
+
+        vscode.window.setStatusBarMessage(
+            localize(
+                'AWS.samcli.deploy.statusbar.message',
+                '$(cloud-upload) Deploying SAM Application to {0}...',
+                deployWizardResponse.stackName,
+            ),
+            deployApplicationPromise
+        )
+    } catch (err) {
+        const error = err as Error
+        channelLogger.logger.error(error)
+
+        channelLogger.channel.show(true)
+        channelLogger.error(
+            'AWS.samcli.deploy.general.error',
+            'An error occurred while deploying a SAM Application. Check the logs for more information.'
+        )
     }
+}
+
+function getBuildRootFolder(deployRootFolder: string): string {
+    return path.join(deployRootFolder, 'build')
+}
+
+function getBuildTemplatePath(deployRootFolder: string): string {
+    return path.join(getBuildRootFolder(deployRootFolder), 'template.yaml')
+}
+
+function getPackageTemplatePath(deployRootFolder: string): string {
+    return path.join(deployRootFolder, 'template.yaml')
+}
 
-    const { region, template, s3Bucket, stackName, parameterOverrides } = args
-    const deployApplicationPromise = (async () => {
-        const tempFolder = await makeTemporaryToolkitFolder('samDeploy')
-        const buildDestination = path.join(tempFolder, 'build')
-        const buildTemplatePath = path.join(buildDestination, 'template.yaml')
-        const outputTemplatePath = path.join(tempFolder, 'template.yaml')
-        let stage: 'starting up' | 'building' | 'packaging' | 'deploying' = 'starting up'
+async function buildPackageDeploy(params: {
+    deployParameters: DeploySamApplicationParameters,
+    invoker: SamCliProcessInvoker,
+    channelLogger: ChannelLogger,
+}): Promise<void> {
+    let stage: 'starting up' | 'building' | 'packaging' | 'deploying' = 'starting up'
```

Comment:
whilst this pattern is pretty neat - it feels to me like because this is required the method might be doing too much - might make more sense to split into more manageable pieces?

Reply:
That's exactly what I'm working up to in a follow-up PR.
(this PR largely rearranged existing code)